### PR TITLE
Refactor AEAD for new MPC API

### PR DESCRIPTION
**Pull request by sinui0**
This PR refactors our `aead` crate to use the new MPC API.

# Changes
- Docs, lints, clippy :paperclip: 
- Unified the interface into a single trait
- Uses the new MPC API, this mainly means the AEAD no longer deals with the encoder, labels, or using OTPs for different privacy configs.
- Tweaked our `universal-hash` crate a bit to make `UniversalHash` boxable. To do this I ditched the associated consts, which weren't being used anyhow, and added a `Send` bound.
- Switched to `Box<dyn Trait>` instead of generics for a simpler type signature.

### CODE REVIEW

File: aead/src/aes_gcm/mod.rs [hunk 1]
```diff
@@ -1,91 +1,324 @@
...
-            ));
+/// An implementation of 2PC AES-GCM.
+pub struct MpcAesGcm {
+    config: AesGcmConfig,
+
+    channel: AeadChannel,
+
+    aes_block: Box<dyn BlockCipher<Aes128>>,
+    aes_ctr: Box<dyn StreamCipher<Aes128Ctr>>,
+    ghash: Box<dyn UniversalHash>,
```
Comment (th4s) [inline aead/src/aes_gcm/mod.rs:36]: :rainbow: very nice

**Review by th4s — APPROVED** [review]
:rocket: ack

File: aead/src/aes_gcm/tag.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    // pad data to be a multiple of 16 bytes
+    let aad_padded_block_count = (aad.len() / 16) + (aad.len() % 16 != 0) as usize;
+    aad.resize(aad_padded_block_count * 16, 0);
+
+    let ciphertext_padded_block_count =
+        (ciphertext.len() / 16) + (ciphertext.len() % 16 != 0) as usize;
+    ciphertext.resize(ciphertext_padded_block_count * 16, 0);
+
+    let mut data: Vec<u8> = Vec::with_capacity(aad.len() + ciphertext.len() + 8);
```
Comment (themighty1) [inline aead/src/aes_gcm/tag.rs:59]: shouldn't it be +16 at the end instead of +8 ?

File: aead/src/lib.rs [hunk 3]
```diff
@@ -55,154 +73,77 @@ pub trait AeadLeader {
     /// * `explicit_nonce` - The explicit nonce to use for encryption.
     /// * `plaintext` - The plaintext to encrypt.
     /// * `aad` - Optional additional authenticated data.
```
Comment (themighty1) [inline aead/src/lib.rs:75]: since aad is not an `Option<>`, we could remove the word optional from the comments here and below.

**Pull request merged**